src/qibolab/instruments/rfsoc.py

```diff
@@ -0,0 +1,488 @@
+""" RFSoC fpga driver.
+
+Supports the following FPGA:
+    RFSoC 4x2
+"""
+import numpy as np
+from qick import AveragerProgram, QickSoc
+
+from qibolab.instruments.abstract import AbstractInstrument, InstrumentException
+from qibolab.pulses import (
+    Drag,
+    Gaussian,
+    Pulse,
+    PulseSequence,
+    PulseShape,
+    PulseType,
+    ReadoutPulse,
+    Rectangular,
+)
+from qibolab.result import ExecutionResults
+from qibolab.sweeper import Parameter, Sweeper
+
+
+class ExecutePulseSequence(AveragerProgram):
+    """This qick AveragerProgram handles a qibo sequence of pulse"""
+
+    def __init__(self, soc, cfg, sequence):
+        """In this function the most important settings are defined and the sequence is transpiled.
+
+        * set the conversion coefficients to be used for frequency and time values
+        * max_gain, adc_trig_offset, max_sampling_rate are imported from cfg (runcard settings)
+        * connections are defined (drive and readout channel for each qubit)
+
+        * pulse_sequence, readouts, channels are defined to be filled by convert_sequence()
+
+        * cfg["reps"] is set from hardware_avg
+        * super.__init__
+        """
+
+        # conversion coefficients
+        self.MHz = 0.000001
+        self.mu_s = 0.001
+
+        # settings
+        self.max_gain = cfg["max_gain"]  # TODO redundancy
+        self.adc_trig_offset = cfg["adc_trig_offset"]
+        self.max_sampling_rate = cfg["sampling_rate"]
+        self.relax_delay = cfg["repetition_duration"]
+        self.syncdelay = 200  # TODO maybe better in runcard
+        cfg["reps"] = cfg["nshots"]
+
+        # connections  (for every qubit here are defined drive and readout lines)
+        self.connections = {
+            "0": {"drive": 1, "readout": 0, "adc_ch": 0},
+        }
+
+        self.pulse_sequence = {}
+        self.readouts = []
+        self.channels = []
+
+        # fill the self.pulse_sequence and the self.readout_pulses oject
+        self.soc = soc
+        self.soccfg = soc  # No need for a different soc config object since qick is on board
+        self.convert_sequence(sequence)
+
+        super().__init__(soc, cfg)
+
+    def convert_sequence(self, sequence):
```

Comment:
I am still not sure why you need this intermediate representation. I think you can use the qibolab `PulseSequence` directly in `add_pulse_to_register`.
If you really need an intermediate representation of the `Pulse` object that is closer to what qick needs, I would suggest to create a seperate `QickPulse` class (or dataclass), instead of using a dictionary. We have done something similar for other instruments (QM, Zurich).

Reply:
This is solved now and everything is done directly in `add_pulse_to_register`